janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphEdgeVertexStep.java

```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop.optimize;
+
+import org.janusgraph.core.JanusGraphMultiVertexQuery;
+
+import com.google.common.collect.Sets;
+
+import org.apache.tinkerpop.gremlin.process.traversal.Traverser;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.EdgeVertexStep;
+import org.apache.tinkerpop.gremlin.process.traversal.util.FastNoSuchElementException;
+import org.apache.tinkerpop.gremlin.structure.Direction;
+import org.apache.tinkerpop.gremlin.structure.Edge;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A JanusGraphEdgeVertexStep is identical to a {@link EdgeVertexStep}. The only difference
+ * being that it can use multiQuery to pre-fetch the vertex properties prior to the execution
+ * of any subsequent has steps and so eliminate the need for a network trip for each vertex.
+ */
+public class JanusGraphEdgeVertexStep extends EdgeVertexStep implements MultiQueriable<Edge, Vertex> {
+
+    private boolean initialized = false;
+    private boolean useMultiQuery = false;
+    private final int txVertexCacheSize;
+
+    public JanusGraphEdgeVertexStep(EdgeVertexStep originalStep, int txVertexCacheSize) {
+        super(originalStep.getTraversal(), originalStep.getDirection());
+        originalStep.getLabels().forEach(this::addLabel);
+        this.txVertexCacheSize = txVertexCacheSize;
+    }
+
+    @Override
+    public void setUseMultiQuery(boolean useMultiQuery) {
+        this.useMultiQuery = useMultiQuery;
+    }
+
+    private void initialize() {
+        assert !initialized;
+        initialized = true;
+
+        if (useMultiQuery) {
+            if (!starts.hasNext()) {
+                throw FastNoSuchElementException.instance();
+            }
+            List<Traverser.Admin<Edge>> edges = new ArrayList<>();
+            Set<Vertex> vertices = Sets.newHashSet();
+            starts.forEachRemaining(e -> {
+                edges.add(e);
+
+                if (vertices.size() < txVertexCacheSize) {
```

Comment:
That's right, and following on from your earlier comment about What happens to a traversal that has a limit() it might be possible to limit it further based on that. I'll investigate.